### PR TITLE
ENH: Allow ``where`` argument to override ``__array_ufunc__``

### PR DESCRIPTION
**Pull request by roytsmart**
Currently, `numpy.ndarray.__array_ufunc__` does not allow the `where` argument to override the default behavior of the ufunc.

This pull request allows this behavior by extending `PyUFunc_CheckOverride` to accept an additional argument that represents `where`, and checks to see if it overrides `__array_ufunc__`.

Closes Issue #23219

### CODE REVIEW

**Comment by byrdie** [comment]
Does anyone know why the Travis CI build is failing? I've tried making sure my branch is ahead of numpy:main, but it still seems to be timing out.

**Review by mhvk — APPROVED** [review]
Wow, I'm impressed with our original work and in particular @seberg's refactor that this is now so easy! It looks all good to me, with the travis failure surely a red herring. I'll approve but not merge yet, so @seberg can have a look too.

**Comment by mhvk** [comment]
Not sure why I always think of things after sending in a review, but what is still needed is a changelog fragment. And relevant updates to the documentation (and an update/clarification of the NEP??).

**Comment by seberg** [comment]
Agreed, a brief compat release note and modifying the NEP (I guess it is the main doc we have) would be good.  Would also be nice to check up on what some `__array_functions__` do that may be in a similar position (e.g. copyto maybe).

I was briefly considering if we should refactor out the `append_if_new` part of the loop into a function rather than the 3 way if, but probably not really clearer.

So I just agree with Marten, the code and test look nice and clean :).

File: doc/neps/nep-0013-ufunc-overrides.rst
Comment (mhvk) [inline doc/neps/nep-0013-ufunc-overrides.rst:176]: looks good but I think it is probably good to mention somewhere (on top?) that this adjustment has been made. Not quite sure what the proper protocol for that is! @seberg?

Comment (byrdie) [reply doc/neps/nep-0013-ufunc-overrides.rst:176]: Agreed, the fact that this NEP says "Status: Final" at the top says to me that this document is sort of supposed to be immutable right?

File: numpy/core/tests/test_overrides.py
Comment (byrdie) [inline numpy/core/tests/test_overrides.py:159]: What do we think about only testing `numpy.negative` here? Is that sufficient or is there a better way to test "all" ufuncs.

File: numpy/core/tests/test_overrides.py
Comment (byrdie) [inline numpy/core/tests/test_overrides.py:273]: @sberg, I used `std` instead of `copyto` since I thought it would be simpler with only one argument. Is this ok? I can change it to `copyto` if you think that would be better. Also, do we think I put this test in the right class? I kept going back and forth on where to put this.

File: doc/neps/nep-0013-ufunc-overrides.rst
Comment (mhvk) [reply doc/neps/nep-0013-ufunc-overrides.rst]: Indeed, but I think this counts as a "clarification" -- in the end, we want people to be able to look at the NEP and be able to get how things work, and why things have changed. In astropy, we use a `data-last-revised` for that purpose -- see https://github.com/astropy/astropy-APEs#updating-apes

But numpy seems less clear. Two possibilities:
1. Add yourself as `author` and then insert the present date; I think this should be below "status: final" so that it is clear you made a small change.
2. Just add something like
```
.. note: This NEP was amended by <you> on <present date> to reflect the decision to allow 
       override also by `where` arguments.
```
I'd tend to go with (2) as being clearer...

Comment (byrdie) [reply doc/neps/nep-0013-ufunc-overrides.rst]: @mhvk, I looked through all the NEPs to see what we've done in the past and it seems like the closest thing was the `:Updated:` tag, so added that with myself as an author below the Status:Final line.

File: numpy/core/tests/test_overrides.py
Comment (mhvk) [reply numpy/core/tests/test_overrides.py:159]: I think it is good enough, though perhaps it is a good idea to verify that its override is considered last? I.e., put in another override in either input or output and check that those have precedence...

**Review by mhvk — COMMENTED** [review]
One nitpick and one suggestion, but really this looks great!

File: doc/release/upcoming_changes/23240.compatibility.rst
Comment (byrdie) [reply doc/release/upcoming_changes/23240.compatibility.rst]: I wrote "Array-likes that define ``__array_ufunc__`` can now override ufuncs if used as the ``where`` keyword argument", how does that sound?

Comment (mhvk) [reply doc/release/upcoming_changes/23240.compatibility.rst]: Sounds good to me!

File: numpy/core/tests/test_overrides.py
Comment (seberg) [reply numpy/core/tests/test_overrides.py:280]: Sure, was just trying to make sure that we handle `where=` similar in both overrdies.

File: doc/release/upcoming_changes/23240.compatibility.rst
Comment (seberg) [inline doc/release/upcoming_changes/23240.compatibility.rst:2]: Would love to shorten, but seems fine.  I think some parsers might not like the underline to have a different length (but I am not sure).

File: doc/release/upcoming_changes/23240.compatibility.rst
Comment (seberg) [inline doc/release/upcoming_changes/23240.compatibility.rst:6]: So one thing here is that objects like dask or masked arrays may need to unwrap `where=` if it exists.  I suspect `dask` already does (or doesn't even do that anymore).

(I.e. objects doing recursive calls will have to make sure that they also clean up `where=` if it exists.  I suspect all others shouldn't care much, although end-users could.  I doubt that is a problem, but wonder if we should mention it here).

In theory, an object (say dask) could even decide that if it was only passed as `where=` it should just return a NumPy array instead.  In practice, I doubt anyone would bother doing such complicated logic.

In any case, I am happy if Marten is, but we should give e.g. @shoyer (and maybe other stake-holders) a chance to comment.  Maybe even just ping the list briefly.

Comment (mhvk) [reply doc/release/upcoming_changes/23240.compatibility.rst:6]: Good point that we should work in the compatibility note that `where` now has to be unwrapped. I don't do this in astropy and may try to quicky run a test with it to see if we would get hit by this.

File: doc/release/upcoming_changes/23240.compatibility.rst
Comment (byrdie) [reply doc/release/upcoming_changes/23240.compatibility.rst]: Ok, I shortened it to just "Array-likes that define ``__array_ufunc__`` can now override ufuncs if used as ``where``" since I think it's pretty obvious that we're talking about the keyword argument.

File: doc/release/upcoming_changes/23240.compatibility.rst
Comment (byrdie) [reply doc/release/upcoming_changes/23240.compatibility.rst:6]: @seberg, good catch on how this will behave a little differently for downstream users. I've added a sentence to the compatibility notes to try and capture what you've explained here.

Comment (byrdie) [reply doc/release/upcoming_changes/23240.compatibility.rst:6]: @mhvk, I just ran the astropy tests against this PR and I did get one failure in `astropy.utils.masked.tests.test_function_helpers.TestUfuncLike.test_fix`, not quite sure how to resolve it yet.
```python3
======================================================= FAILURES =======================================================
________________________________________________ TestUfuncLike.test_fix ________________________________________________

self = <astropy.utils.masked.tests.test_function_helpers.TestUfuncLike object at 0x7f260513df90>

    def test_fix(self):
>       self.check(np.fix)

astropy/utils/masked/tests/test_function_helpers.py:644:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
astropy/utils/masked/tests/test_function_helpers.py:68: in check
    o = func(self.ma, *args, **kwargs)
astropy/utils/masked/core.py:841: in __array_function__
    return super().__array_function__(function, types, args, kwargs)
../.local/lib/python3.10/site-packages/numpy/lib/ufunclike.py:53: in func
    return f(x, out=out, **kwargs)
../.local/lib/python3.10/site-packages/numpy/lib/ufunclike.py:118: in fix
    res = nx.floor(x, out=res, where=nx.greater_equal(x, 0))
astropy/utils/masked/core.py:827: in __array_ufunc__
    result = getattr(ufunc, method)(*unmasked, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = MaskedNDArray([[  ———,  True,  True],
               [ True,   ———,  True]]), ufunc = <ufunc 'floor'>
method = '__call__', inputs = (array([[0., 1., 2.],
       [3., 4., 5.]]),)
kwargs = {'where': MaskedNDArray([[  ———,  True,  True],
               [ True,   ———,  True]])}
out = (array([[0., 1., 2.],
       [3., 4., 5.]]),), out_unmasked = (array([[0., 1., 2.],
       [3., 4., 5.]]),)
out_mask = None, out_masks = (None,), d = array([[0., 1., 2.],
       [3., 4., 5.]]), m = None

    def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
        out = kwargs.pop("out", None)
        out_unmasked = None
        out_mask = None
        if out is not None:
            out_unmasked, out_masks = self._get_data_and_masks(*out)
            for d, m in zip(out_unmasked, out_masks):
                if m is None:
                    # TODO: allow writing to unmasked output if nothing is masked?
                    if d is not None:
>                       raise TypeError("cannot write to unmasked output")
E                       TypeError: cannot write to unmasked output

astropy/utils/masked/core.py:700: TypeError
```

Comment (mhvk) [reply doc/release/upcoming_changes/23240.compatibility.rst:6]: `np.fix` was one of the numpy functions that worked directly on the `Masked` class -- but clearly its implementation used `where`. That will be fairly trivial to fix.

So, I don't think that is a show-stopper, but does show that @seberg is correct that not unwrapping `where` arguments is likely the main issue downstream may have. That said, it should mostly be a problem for subclassing `ndarray`.

Comment (byrdie) [reply doc/release/upcoming_changes/23240.compatibility.rst:6]: Ya, this is a good demonstration of what can happen in other downstream packages. I'm happy to open a PR in astropy to fix this when the time comes, but hopefully others will chime in about how they think this will affect other packages.

Comment (mhvk) [reply doc/release/upcoming_changes/23240.compatibility.rst:6]: Thanks in advance! We're doing overrides now for `Quantity`, `Masked`, and `Distribution`, but the code is fairly similar.

p.s. Inspired by your original use-case, I got at least a little further in updating our `Distribution`... see https://github.com/astropy/astropy/pull/14421 and https://github.com/astropy/astropy/pull/14423

Comment (seberg) [reply doc/release/upcoming_changes/23240.compatibility.rst:6]: Ah, while I may have been right that in some cases unwrapping is necessary. I had a brief look at `dask` and I think it should be fine.  Of course no existing implementation will defer correctly, because unlike `__array_function__` we do not pass all types/objects explicitly.

That means that also here `ndarray.__array_function__` is very much missing a way to defer to the object passed in by `where=`.

The tests work here, but I guess that is just due to the "subclasses before superclasses" rule (did not double check).

So to actually take full effect, all downstream projects need to start checking `where=`  (this makes me wish we had that explicit list like `__array_function__` does...).

Comment (byrdie) [reply doc/release/upcoming_changes/23240.compatibility.rst:6]: > That means that also here `ndarray.__array_function__` is very much missing a way to defer to the object passed in by `where=`.
Sorry, I did not follow your logic here, can you explain it a little more? I'm confused since `DuckArray` in `TestImplementingArgs.test_where_duck_array` is not a subclass of `numpy.ndarray`, so I don't see why the "subclasses before superclasses" rule applies.

Comment (seberg) [reply doc/release/upcoming_changes/23240.compatibility.rst:6]: Ah, `__array_ufunc__` works a bit different, since it rarely does a recursive call I think, and using `super().__array_ufunc__` is dubious.
In any case, I think we should change it, although I am not sure how to best test it.

An `__array_ufunc__` implementation normally does something like:
```
def __array_ufunc__(self, args, **kwargs):
    objs = args + kwargs.get(out, ())
    for obj in objs:
        if not is_of_known_type(obj):
             reutn NotImplemented
```
Which arguably should now be:
```
def __array_ufunc__(self, args, **kwargs):
    objs = args + kwargs.get(out, ())
    if "where" in kwargs:
        objs.append(kwargs.get("where"))
```
(or something more pretty, I don't love it, but probably its good)

In any case, we have an `__array_function__` implementation in NumPy and maybe an example in the NEP and both should probably use this pattern?

Comment (mhvk) [reply doc/release/upcoming_changes/23240.compatibility.rst:6]: Let's be sure not to confuse `__array_ufunc__` and `__array_function__` - for the last one, there was a systematic go-through of all functions to see what arguments could be arrays.

But @seberg does have a good point: what I think is missing is for `ndarray.__array_ufunc__` to check whether `where` is present, and, like for the inputs and outputs, return `NotImplemented` as needed. (Since astropy typically `super()`s to `ndarray`, that might actually expose a few more problems, but we'll see...). `ndarray.__array_ufunc__` is implemented in `src/multiarray/methods.c`.

Comment (mhvk) [reply doc/release/upcoming_changes/23240.compatibility.rst:6]: p.s. Sorry for the duplication, @seberg! But at least I hopefully saved @byrdie from having to find where `ndarary.__array_ufunc__` is defined.  And good point about adjusting examples!

Comment (byrdie) [reply doc/release/upcoming_changes/23240.compatibility.rst:6]: Huh, I guess I'm confused why the tests seem to be working now. Do you know of a test I can write that would fail the way the code is now?

Comment (seberg) [reply doc/release/upcoming_changes/23240.compatibility.rst:6]: You may have to use `super().__array_function__` in a subclass.  (And that may always be a bit dubious anyway, see gh-23014... So, in a sense, it maybe doesn't really matter much.

Comment (mhvk) [reply doc/release/upcoming_changes/23240.compatibility.rst:6]: I think we need a test for an ndarray subclass that overrides `__array_ufunc__`  and by mistake passes its own instance down as where, like,
```
class Mine(np.ndarray):
    def __array_ufunc__(...):
        ...
        return super().__array_ufunc__(..., where=<instance-of-Mine>)
```

Maybe the best place is (near) `test_ufunc_override_with_super` in `test_umath.py`?

Comment (byrdie) [reply doc/release/upcoming_changes/23240.compatibility.rst:6]: @mhvk, I'm trying to implement a test like you suggested above, but it's leading to infinite recursion since `Mine.__array_ufunc__` is basically a no-op, and then `ndarray.__array_ufunc__` sees the instance of `Mine` and calls `Mine.__array_ufunc__` again. Is this what you expected to happen?

Comment (mhvk) [reply doc/release/upcoming_changes/23240.compatibility.rst:6]: You are getting the recursion before any change in `ndarray.__array_ufunc__`, right? If so, that's not what I initially expected, but good in the sense that it means nobody can have made the mistake of passing in their subclass to `where` and not notice!

But with a change in `ndarray.__array_ufunc__`, that should return `NotImplemented` and then the result should be a `TypeError` (which is what the test would check).

Comment (byrdie) [reply doc/release/upcoming_changes/23240.compatibility.rst:6]: Yeah, no change in `ndarray.__array_ufunc__`. Sounds good though, thanks for walking me through this!

Comment (byrdie) [reply doc/release/upcoming_changes/23240.compatibility.rst:6]: @seberg and @mhvk, I think I've sorted this out. I've modified `ndarray.__array_ufunc__` to return `NotImplemented` if `where` overrides `__array_ufunc__`, added a test along the lines suggested by @mhvk, and updated the NEP to reflect this pattern. Let me know what you think.

File: numpy/core/src/multiarray/methods.c
Comment (mhvk) [inline numpy/core/src/multiarray/methods.c:1137]: I worry a bit about speed, but `out` is treated very similarly so definitely out of scope here. But maybe @seberg has thoughts...

**Review by mhvk — APPROVED** [review]
Looks really good! Only a suggestion for a textual change left. Re-approving, though probably @seberg should have a last look too.

File: numpy/core/src/multiarray/methods.c
Comment (seberg) [reply numpy/core/src/multiarray/methods.c:1137]: This can be much faster by using a global `where` string rather than `"string"`.  We can't avoid the ~20ns of the dict lookup itself, unfortunately.  Now that we go beyond one key that needs to be fetched from the dict, it could have been nice to do that earlier.

That said, we already have to go through all of the work to build the dict, so that is more work anyway :).

Comment (byrdie) [reply numpy/core/src/multiarray/methods.c:1137]: So I should go ahead and define a global `where` string right?

Comment (seberg) [reply numpy/core/src/multiarray/methods.c:1137]: Yeah, i.e. there: https://github.com/numpy/numpy/blob/main/numpy/core/src/multiarray/multiarraymodule.c#L4901
It probably already is in the umath namespace (which is the same one now).  But could put it there (and at some point we could move the umath ones over maybe).

Comment (byrdie) [reply numpy/core/src/multiarray/methods.c]: > Now that we go beyond one key that needs to be fetched from the dict, it could have been nice to do that earlier.

Sorry, I don't understand, what could we do earlier?

Comment (seberg) [reply numpy/core/src/multiarray/methods.c]: Ignore it.  `__array_function__` passes all objects that need dispatching explicitly (a second time).  Which would be looking nice now that we have 3 things where objects can be found.

Comment (byrdie) [reply numpy/core/src/multiarray/methods.c]: @seberg, I'm having trouble accessing the new global variable I've created. When I include `multiarraymodule.h` in `methods.c`, I get [linker errors](https://github.com/numpy/numpy/actions/runs/4260484958/jobs/7413700120#step:5:592). Any ideas on where I might be going wrong?

Comment (seberg) [reply numpy/core/src/multiarray/methods.c]: @byrdie did you see how the other symbols there were defined used?  You probably forgot the `extern`.

Comment (byrdie) [reply numpy/core/src/multiarray/methods.c]: Thanks, it ended up being something else, but that was a good guess! I usually forget the extern. Should be all good to go as soon as the CI passes.

File: numpy/core/src/multiarray/methods.c
Comment (byrdie) [inline numpy/core/src/multiarray/methods.c:31]: @seberg Is this the appropriate place to include this header?

**Comment by mhvk** [comment]
Not relevant for this PR, but `"out"` is gotten from the dictionary in the same way as "where" was here. @seberg - would it be worth raising a new issue with possible speed-ups? I might be able to do it in a bit, but am not sure I understood your other suggestion.

**Comment by seberg** [comment]
> Not relevant for this PR, but "out" is gotten from the dictionary in the same way as "where" was here

Oh sorry, in that case there was no reason to update it really...  This path is not hot anyway (I suspect few users use `super().__array_function__`).

No other suggestion beyond using interned strings.  array-function passes `types`, which means you can decide whether to defer without ever looking at the exact arguments.  I doubt we want to go into looking into an iteration of how dispatching works here though.

**Comment by byrdie** [comment]
@seberg, if we wanted to defer using `types`, wouldn't we have to change the signature of `__array_ufunc__`?

**Comment by mhvk** [comment]
@byrdie - indeed, which would be hard to do...  Might need some opt-in signal or so.

**Comment by byrdie** [comment]
Yeah, I think that's probably out of scope for this PR, though I agree it would probably be better/faster if we did do that!

**Comment by byrdie** [comment]
@seberg, what can I do to get this merged?

**Comment by seberg** [comment]
Considering that nobody voiced concerns, I think it is time to give this a shot.  If anyone has concerns we could pull it out (at least easily before the next final release).

Thanks @byrdie.

**Comment by byrdie** [comment]
Thanks @seberg and @mhvk!

**Pull request merged**
